File: ollama/codigo-final/codegemma-7b/cot/temp_0.7/nbody.cpp
#include <iostream>
#include <iomanip>

using namespace std;

const double PI = 3.141592653589793;
const double SOLAR_MASS = 4 * PI * PI;
const double DAYS_PER_YEAR = 365.24;

struct Body {
  double x, y, z;
  double vx, vy, vz;
  double mass;
};

double distance(Body& a, Body& b) {
  return sqrt((a.x - b.x) * (a.x - b.x) + (a.y - b.y) * (a.y - b.y) + (a.z - b.z) * (a.z - b.z));
}

double gravitationalForce(Body& a, Body& b) {
  double d = distance(a, b);
  double G = 6.67430e-11;
  return G * a.mass * b.mass / (d * d);
}

double kineticEnergy(Body& b) {
  return 0.5 * b.mass * (b.vx * b.vx + b.vy * b.vy + b.vz * b.vz);
}

double potentialEnergy(Body& a, Body& b) {
  return -gravitationalForce(a, b) * distance(a, b);
}

double totalEnergy(Body& sun, Body& jupiter, Body& saturn, Body& uranus, Body& neptune) {
  return kineticEnergy(sun) + potentialEnergy(sun, jupiter) + potentialEnergy(sun, saturn) + potentialEnergy(sun, uranus) + potentialEnergy(sun, neptune) +
         kineticEnergy(jupiter) + kineticEnergy(saturn) + kineticEnergy(uranus) + kineticEnergy(neptune);
}

void updateVelocities(Body& a, Body& b) {
  double f = gravitationalForce(a, b);
  double d = distance(a, b);
  a.vx -= f * (a.x - b.x) / (d * d * a.mass);
  a.vy -= f * (a.y - b.y) / (d * d * a.mass);
  a.vz -= f * (a.z - b.z) / (d * d * a.mass);
  b.vx += f * (a.x - b.x) / (d * d * b.mass);
  b.vy += f * (a.y - b.y) / (d * d * b.mass);
  b.vz += f * (a.z - b.z) / (d * d * b.mass);
}

void updatePositions(Body& b) {
  b.x += b.vx * DAYS_PER_YEAR;
  b.y += b.vy * DAYS_PER_YEAR;
  b.z += b.vz * DAYS_PER_YEAR;
}

int main() {
  int N;
  cin >> N;

  Body sun = {0, 0, 0, 0, 0, 0, SOLAR_MASS};
  Body jupiter = {...};
  Body saturn = {...};
  Body uranus = {...};
  Body neptune = {...};

  double offsetMomentum = (jupiter.vx + saturn.vx + uranus.vx + neptune.vx) / -sun.mass;
  sun.vx += offsetMomentum;

  for (int t = 0; t < N; t++) {
    updateVelocities(sun, jupiter);
    updateVelocities(sun, saturn);
    updateVelocities(sun, uranus);
    updateVelocities(sun, neptune);

    updateVelocities(jupiter, saturn);
    updateVelocities(jupiter, uranus);
    updateVelocities(jupiter, neptune);

    updateVelocities(saturn, uranus);
    updateVelocities(saturn, neptune);

    updateVelocities(uranus, neptune);

    updatePositions(sun);
    updatePositions(jupiter);
    updatePositions(saturn);
    updatePositions(uranus);
    updatePositions(neptune);
  }

  double initialEnergy = totalEnergy(sun, jupiter, saturn, uranus, neptune);
  double finalEnergy = totalEnergy(sun, jupiter, saturn, uranus, neptune);

  cout << fixed << setprecision(9) << initialEnergy * 100 / finalEnergy << endl;

  return 0;
}

File: ollama/codigo-final/nbody/qwen2.5-coder-7b/cot/temp_0.7/nbody.py
import sys
import numpy as np

# Constants
PI = 3.141592653589793
SOLAR_MASS = 4 * PI * PI
DAYS_PER_YEAR = 365.24

# Initial conditions for the planets
initial_conditions = {
    'Sun': {'x': 0, 'y': 0, 'z': 0, 'vx': 0, 'vy': 0, 'vz': 0, 'mass': SOLAR_MASS},
    'Jupiter': {'x': 4.84143144246472090e+00, 'y': -1.16032004402742839e+00, 'z': -1.03622044471123109e-01,
                'vx': 1.66007664274403694e-03 * DAYS_PER_YEAR, 'vy': 7.69901118419740425e-03 * DAYS_PER_YEAR,
                'vz': -6.90460016972063023e-05 * DAYS_PER_YEAR, 'mass': 9.54791938424326609e-04 * SOLAR_MASS},
    'Saturn': {'x': 8.34336671824457987e+00, 'y': 4.12479856412430479e+00, 'z': -4.03523417114321381e-01,
               'vx': -2.76742510726862411e-03 * DAYS_PER_YEAR, 'vy': 4.99852801234917238e-03 * DAYS_PER_YEAR,
               'vz': 2.30417297573763929e-05 * DAYS_PER_YEAR, 'mass': 2.85885980666130812e-04 * SOLAR_MASS},
    'Uranus': {'x': 1.28943695621391310e+01, 'y': -1.51111514016986312e+01, 'z': -2.23307578892655734e-01,
               'vx': 2.96460137564761618e-03 * DAYS_PER_YEAR, 'vy': 2.37847173959480950e-03 * DAYS_PER_YEAR,
               'vz': -2.96589568540237556e-05 * DAYS_PER_YEAR, 'mass': 4.36624404335156298e-05 * SOLAR_MASS},
    'Neptune': {'x': 1.53796971148509165e+01, 'y': -2.59193146099879641e+01, 'z': 1.79258772950371181e-01,
                'vx': 2.68067772490389322e-03 * DAYS_PER_YEAR, 'vy': 1.62824170038242295e-03 * DAYS_PER_YEAR,
                'vz': -9.51592254519715870e-05 * DAYS_PER_YEAR, 'mass': 5.15138902046611451e-05 * SOLAR_MASS}
}

# Function to calculate gravitational force
def gravitational_force(masses, positions):
    N = len(masses)
    F = np.zeros((N, 3))
    for i in range(N):
        for j in range(i+1, N):
            dx = positions[j][0] - positions[i][0]
            dy = positions[j][1] - positions[i][1]
            dz = positions[j][2] - positions[i][2]
            r_squared = dx**2 + dy**2 + dz**2
            Fij = SOLAR_MASS * masses[i] * masses[j] / (r_squared * np.sqrt(r_squared))
            F[i] += Fij * (dx / r_squared, dy / r_squared, dz / r_squared)
            F[j] -= Fij * (dx / r_squared, dy / r_squared, dz / r_squared)
    return F

# Function to update positions and velocities using a symplectic integrator
def symplectic_integrator(masses, positions, velocities, dt):
    N = len(masses)
    F = gravitational_force(masses, positions)
    
    # Update velocities first
    for i in range(N):
        velocities[i] += F[i] * dt / masses[i]
    
    # Update positions next
    for i in range(N):
        positions[i] += velocities[i] * dt

# Function to calculate total energy of the system
def total_energy(masses, positions, velocities):
    N = len(masses)
    kinetic_energy = 0.5 * np.sum([masses[i] * (vx**2 + vy**2 + vz**2) for i in range(N)])
    potential_energy = 0
    for i in range(N):
        for j in range(i+1, N):
            dx = positions[j][0] - positions[i][0]
            dy = positions[j][1] - positions[i][1]
            dz = positions[j][2] - positions[i][2]
            r_squared = dx**2 + dy**2 + dz**2
            potential_energy -= SOLAR_MASS * masses[i] * masses[j] / np.sqrt(r_squared)
    return kinetic_energy + potential_energy

# Adjust the Sun's initial velocity to zero total momentum
def offsetMomentum(masses, positions, velocities):
    px, py, pz = 0, 0, 0
    for i in range(len(masses)):
        px += masses[i] * velocities[i][0]
        py += masses[i] * velocities[i][1]
        pz += masses[i] * velocities[i][2]
    velocities[0][0], velocities[0][1], velocities[0][2] = -px, -py, -pz

# Main function
def main(N):
    # Initialize positions and velocities
    positions = [np.array([initial_conditions[planet]['x'], initial_conditions[planet]['y'],
                           initial_conditions[planet]['z']]) for planet in initial_conditions]
    velocities = [np.array([initial_conditions[planet]['vx'], initial_conditions[planet]['vy'],
                            initial_conditions[planet]['vz']]) for planet in initial_conditions]
    
    # Adjust Sun's velocity to zero total momentum
    offsetMomentum(initial_conditions['Sun']['mass'], positions, velocities)
    
    # Calculate initial energy
    initial_energy = total_energy([initial_conditions[planet]['mass'] for planet in initial_conditions],
                                 positions, velocities)
    
    # Simulate the system
    dt = 0.01 * DAYS_PER_YEAR / N
    for _ in range(N):
        symplectic_integrator([initial_conditions[planet]['mass'] for planet in initial_conditions],
                               positions, velocities, dt)
    
    # Calculate final energy
    final_energy = total_energy([initial_conditions[planet]['mass'] for planet in initial_conditions],
                                positions, velocities)
    
    # Print the initial and final energy in %.9f format
    print(f"{initial_energy:.9f}")
    print(f"{final_energy:.9f}")

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python script.py <N>")
        sys.exit(1)
    
    N = int(sys.argv[1])
    main(N)
